### PR TITLE
JAVA-2521

### PR DESCRIPTION
**Pull request by jyemin**
These changes will allow future applications to more easily migrate applications that rely on DBObject from DBCollection to MongoCollection<DBObject>, without having to take a dependency on driver-legacy.

### CODE REVIEW

File: bson/build.gradle [hunk 1]
```diff
@@ -27,6 +27,15 @@ clirr {
     failOnErrors = true
 }
 
+sourceSets {
```
Comment (jyemin) [inline bson/build.gradle]: This may be a mistake to only do this srcDir split for bson, and not driver-core.  There are a lot more classes moving into bson, which made it more compelling there.

File: bson/src/main-legacy/org/bson/BasicBSONEncoder.java [hunk 2]
```diff
@@ -190,15 +189,6 @@ protected void _putObjectField(final String name, final Object initialValue) {
             putCodeWScope(name, (CodeWScope) value);
         } else if (value instanceof Code) {
             putCode(name, (Code) value);
-        } else if (value instanceof DBRef) {
```
Comment (jyemin) [inline bson/src/main-legacy/org/bson/BasicBSONEncoder.java]: This can't be here any more since DBRef isn't accessible from bson.  The upside though is that the DefaultDBEncoder subclass handles DBRef in its putSpecial method.

File: bson/src/main-legacy/org/bson/BasicBSONObject.java [hunk 3]
```diff
@@ -335,17 +335,6 @@ public BasicBSONObject append(final String key, final Object val) {
...
 
-    /**
-     * Returns a JSON serialization of this object
-     *
-     * @return JSON serialization
-     */
-    @Override
-    @SuppressWarnings("deprecation")
-    public String toString() {
-        return com.mongodb.util.JSONSerializers.getStrict().serialize(this);
```
Comment (jyemin) [inline bson/src/main-legacy/org/bson/BasicBSONObject.java]: Also not accessible, nor is a codec for this class (DBObjectCodec moved to core, but there is no BSONObjectCodec).  I think this is justified though as use of this class is rare and it's only a toString method.

File: bson/src/main-legacy/org/bson/BasicBSONObject.java [hunk 4]
```diff
@@ -368,7 +357,7 @@ public boolean equals(final Object o) {
             return false;
         }
 
-        return Arrays.equals(canonicalizeBSONObject(this).encode(), canonicalizeBSONObject(other).encode());
+        return Arrays.equals(getEncoder().encode(canonicalizeBSONObject(this)), getEncoder().encode(canonicalizeBSONObject(other)));
```
Comment (jyemin) [inline bson/src/main-legacy/org/bson/BasicBSONObject.java]: I could revert this change now.  It's leftover from a refactoring that proved unnecessary.

File: bson/src/test/unit-legacy/org/bson/BSONTest.java [hunk 5]
```diff
@@ -55,57 +53,68 @@ public void tearDown() {
...
-                                          final String expectedHash) throws IOException {
+                                          final String expectedHex) throws IOException {
         // check encoding
         BSONEncoder bsonEncoder = new BasicBSONEncoder();
         OutputBuffer buf = new BasicOutputBuffer();
         bsonEncoder.set(buf);
         bsonEncoder.putObject(toEncodeAndDecode);
         assertEquals(expectedEncodedSize, buf.size());
-        assertEquals(expectedHash, hexMD5(buf.toByteArray()));
+        assertEquals(expectedHex, toHex(buf.toByteArray()));
```
Comment (jyemin) [inline bson/src/test/unit-legacy/org/bson/BSONTest.java]: hexMD5 is no longer accessible.  The MD5 part of this was unnecessary though so I replaced it with just hex.

File: driver-core/src/main/com/mongodb/BasicDBList.java [hunk 6]
```diff
@@ -1,11 +1,11 @@
 /*
- * Copyright (c) 2008-2014 MongoDB, Inc.
+ * Copyright 2018 MongoDB, Inc.
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/BasicDBList.java]: Damnit, IntelliJ!  Will fix.

File: driver-core/src/main/com/mongodb/BasicDBObject.java [hunk 7]
```diff
@@ -58,7 +66,7 @@
      * @mongodb.driver.manual reference/mongodb-extended-json/ MongoDB Extended JSON
      */
     public static BasicDBObject parse(final String json) {
-        return parse(json, getDefaultCodecRegistry().get(BasicDBObject.class));
+        return parse(json, DBObjectCodec.getDefaultRegistry().get(BasicDBObject.class));
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/BasicDBObject.java]: Can't just use new DBObjectCodec() here since that's a Decoder<DBObject> rather than a Decoder<BasicDBObject>

File: driver-core/src/main/com/mongodb/BasicDBObject.java [hunk 8]
```diff
@@ -187,6 +195,37 @@ public String toJson(final JsonWriterSettings writerSettings, final Encoder<Basi
         return writer.getWriter().toString();
     }
 
+    @Override
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/BasicDBObject.java]: Overriding this method, because the alternative is to add a new method to BasicBSONObject

     protected BSONEncoder getEncoder()

and override it here to return an instance of DefaultDBEncoder (which can handle DBRef), which I'd prefer not to pull into driver-core.

File: driver-core/src/main/com/mongodb/DBObjectCodec.java [hunk 9]
```diff
@@ -60,8 +64,10 @@
  */
 @SuppressWarnings("rawtypes")
 public class DBObjectCodec implements CollectibleCodec<DBObject> {
-
     private static final BsonTypeClassMap DEFAULT_BSON_TYPE_CLASS_MAP = createDefaultBsonTypeClassMap();
+    private static final CodecRegistry DEFAULT_REGISTRY =
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/DBObjectCodec.java]: Similar now to DocumentCodec

File: bson/build.gradle (hunk 1 above)
Comment (rozza) [reply bson/build.gradle]: Interesting, we can always split out in the future.

File: bson/src/main-legacy/org/bson/BasicBSONEncoder.java (hunk 2 above)
Comment (rozza) [reply bson/src/main-legacy/org/bson/BasicBSONEncoder.java]: Ack

File: bson/src/main-legacy/org/bson/BasicBSONObject.java (hunk 3 above)
Comment (rozza) [reply bson/src/main-legacy/org/bson/BasicBSONObject.java]: Ack

File: bson/src/main-legacy/org/bson/BasicBSONObject.java (hunk 4 above)
Comment (rozza) [reply bson/src/main-legacy/org/bson/BasicBSONObject.java]: Happy either way.

File: bson/src/test/unit-legacy/org/bson/BSONTest.java (hunk 5 above)
Comment (rozza) [reply bson/src/test/unit-legacy/org/bson/BSONTest.java]: Ack

File: driver-core/src/main/com/mongodb/BasicDBObject.java (hunk 8 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/BasicDBObject.java]: Ack - thats fine.

File: driver-core/src/main/com/mongodb/DBObjectCodec.java (hunk 9 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/DBObjectCodec.java]: :)

**Comment by jyemin** [comment]
@rozza can you check if you can import this branch into IntelliJ?  For me it was getting confused by the multiple source sets in the bson module

**Comment by jyemin** [comment]
Combined the bson source sets since we're unlikely to be removing these classes anytime soon.

File: mongo-java-driver/build.gradle [hunk 10]
```diff
@@ -34,7 +34,8 @@ sourceSets {
             srcDirs = ["$rootDir/driver-legacy/src/main",
                        "$rootDir/driver/src/main",
                        "$rootDir/driver-core/src/main",
-                       "$rootDir/bson/src/main"]
+                       "$rootDir/bson/src/main",
+                       "$rootDir/bson/src/main-legacy"]
```
Comment (rozza) [inline mongo-java-driver/build.gradle]: This can now be removed.
